### PR TITLE
Feat/attribute schema

### LINKED ISSUE #26: [RFE] Create initial artifact schema publishing
**What is the problem you're trying to solve**
A schema needs to be published to the registry when publishing OCI Artifacts. The schema will be used to validate attribute content for a collection.

**Describe the solution you'd like**
Since all annotations in a manifest are seen as strings, a textual serialized format for values is required. JSON schema can be used to validate attribute values and keys. 

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by jpower432**
# Description

This PR adds schema publishing and validation to the CLI. It changes the commands from `build` to `build collection` and `build schema`.  Performing a build collection with a schema address will result in the schema validation against the attribute in the dataset config. This does not implement DefaultContentDeclarations, CommonAttributeMapping, or Algorithms specs. This will be done is later when those concepts are fully designed.

A couple of decisions here that may be up for discussion:
- The schema used is a generated JSON schema from user inputs. Only type checks are being completed initially, but using complex checks is possible through the library since the underlying JSON schema field is exported.
- All keys specified in the schema are required in the collection attributes.
- Schema checks are only done during collection publishing, not during retrieval.

Depends on #66 

Closes #26 

## Type of change

- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [X] This change requires a documentation update

# How Has This Been Tested?

1. Create a schema-configuration file to define attribute keys and types for corresponding collections:   
   ```bash
    kind: SchemaConfiguration
    apiVersion: client.uor-framework.io/v1alpha1
    schema:
      attributeTypes:
        "animal": string
        "size": number
        "color": string
        "habitat": string
        "mammal": boolean
   ```
2. Build and save the schema:
    ```
    uor-client-go build schema schema-config.yaml localhost:5000/myschema:latest
    ```
3. Push the schema to the remote registry:
   ```
   uor-client-go push localhost:5000/myschema:latest
   ```
5. Create a new directory.
6. Add the content to be uploaded in the directory (can be files of any content types).
7. Create a dataset-config.yaml that will fail
    ```bash
    kind: DataSetConfiguration
    apiVersion: client.uor-framework.io/v1alpha1
    collection:
      schemaAddress: "localhost:5000/myschema:latest"
      files:
        - file: "fish.jpg"
          attributes:
            animal: "fish"
            habitat: "ocean"
            size: "small"
            color: "blue"
            ## this is the wrong value type
            mammal: "false"
        - file: "subdir1/dog.jpg"
          attributes:
            animal: "dog"
            habitat: "house"
            size: "medium"
            color: "brown"
            mammal: true
        - file: "*.jpg"
          attributes:
            custom: "customval"
    
    ```
8. Build the collection
    ```
    uor-client-go build collection my-workspace localhost:5000/test/dataset:latest --dsconfig dataset-config.yaml 
   Error: mammal: invalid type: expected: boolean, got: string
    ```
9. Create a dataset-config.yaml that will pass
    ```bash
    kind: DataSetConfiguration
    apiVersion: client.uor-framework.io/v1alpha1
    collection:
      schemaAddress: "localhost:5000/myschema:latest"
      files:
        - file: "fish.jpg"
          attributes:
            animal: "fish"
            habitat: "ocean"
            size: "small"
            color: "blue"
            mammal: false
        - file: "subdir1/dog.jpg"
          attributes:
            animal: "dog"
            habitat: "house"
            size: "medium"
            color: "brown"
            mammal: true
        - file: "*.jpg"
          attributes:
            custom: "customval"
    ```
10. Build the collection
    ```
    uor-client-go build collection my-workspace localhost:5000/test/dataset:latest --dsconfig dataset-config.yaml 
    ```

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Requires updates to the [Getting Started Guide](https://universalreference.io/docs/quick-start/)

### CODE REVIEW

**Comment by afflom** [comment]
First of all, this PR is an incredible milestone for UOR. Thank you for your work here. 

I tested this in a few ways and found that it works as stated. I would prefer optional attributes exposed in the config, but understand that this is already a very large PR. I added an issue to track this here: https://github.com/uor-framework/uor-client-go/issues/74

LGTM

**Review by afflom — APPROVED** [review]

**Review by sabre1041 — CHANGES_REQUESTED** [review]
@jpower432 This is beyond amazing! Awesome work. 

Theres a number of enhancements (mainly around formatting in the README that I'd recommend making). However, I do not want to block this PR from getting in.

Would you be able to correct the one inline comment related to the formatting of the `SchemaConfiguration`. I'd be happy to work on the rest of the formatting as a follow up PR

File: README.md [hunk 1]
```diff
@@ -136,19 +142,226 @@ uor-client-go pull localhost:5000/myartifacts:latest -o my-output-directory --at
...
 
+### Collection Publishing with Schema
+1. Create a schema-configuration file to define attribute keys and types for corresponding collections:
+   Example schema-config.yaml
+   
+   ```bash
+    kind: SchemaConfiguration
+    apiVersion: client.uor-framework.io/v1alpha1
+    schema:
+    attributeTypes:
```
[the remark this answers is not in the document]
Comment (jpower432) [reply README.md]: Fixed. I also found one additional place where the formatting was not right for the `SchemaConfiguration`. That copy-paste error explains the formatting in the testing description :).

**Review by sabre1041 — APPROVED** [review]
LGTM

**Pull request merged**